### PR TITLE
[system-probe] Add process monitoring and USM tagging

### PR DESCRIPTION
**Pull request by hmahmood**
### What does this PR do?

* Adds USM tags for connection processes. Only `DD_ENV`, `DD_SERVICE`, and `DD_VERSION` environment variables are added as `env:`, `service:`, and `version:` respectively.
* `runtime_security_config.event_monitoring.enabled` has been replaced with two new configs, `event_monitoring_config.network_process.enabled`, and `event_monitoring_config.process.enabled`, both `false` by default. Both will turn on the runtime security module
* A new config `event_monitoring_config.network_process.max_tracked_processes`, set to `1024` by default; this is the size of the process LRU cache described below
* process data from the runtime security module is stored in a new LRU cache. Only process data for processes that have the USM environment variables (see above) or have a container ID are stored

### Additional Notes

~10% increase in CPU:
system-probe avg k8s CPU Usage by namespace (autosmoothed) (4)

~11% increase in RSS:
system-probe RSS (avg) (6)

### Describe how to test/QA your changes
* enable network tracer process event monitoring by setting in the system-probe config:
```
event_monitoring_config:
  network_process:
    enabled: true
```
* run system-probe with above config
* query the system-probe for connections with `curl --unix-socket /opt/datadog-agent/run/sysprobe.sock http://unix/connections` (note the unix socket path could be different in your environment; check the system-probe log file for the path); this is to register a client with the system-probe
* in a separate console/terminal on the same machine, run 
```shell
DD_ENV=env DD_SERVICE=service DD_VERSION=version FOO=bar curl https://www.google.com
```
* query the system-probe for connections with `curl --unix-socket /opt/datadog-agent/run/sysprobe.sock http://unix/connections` (note the unix socket path could be different in your environment; check the system-probe log file for the path). You should see the `tags` field in the returned json set to 
```
  "tags": [
    "env:env",
    "version:version",
    "service:service"
  ],
```
on the connection entry for google.com, you should see:
```
      "tags": [
        0,
        1,
        2
      ],
```

### Reviewer's Checklist

- [x] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [x] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [x] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [x] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

File: pkg/network/events/monitor.go [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+	}
+
+}
+
+func (e *eventMonitor) HandleCustomEvent(rule *rules.Rule, event *sprobe.CustomEvent) {
+}
+
+func (e *eventMonitor) RegisterHandler(handler HandlerFunc) {
+	if handler != nil {
+		e.handlers = append(e.handlers, handler)
```
Comment (vickenty) [inline pkg/network/events/monitor.go:96]: Just to double check: does this module support concurrent use or not? The use of `Once` and `atomic.Value` in this module seem to suggest that there is concurrency involved; but modification of the `eventMonitor` happens without synchronization, which I think is needed to avoid races.

Comment (hmahmood) [reply pkg/network/events/monitor.go:96]: The `once` and `atomic.Value` usage is for initialization and safe use of the global monitor object only since it is used and accessed from outside the network tracer. But I think you're right, this should be protected as well; I'll add a lock around this.

File: pkg/network/tracer/process_cache.go [hunk 2]
```diff
@@ -21,28 +24,119 @@ var defaultFilteredEnvs = []string{
...
+	// that a process in the cache must have; empty filteredEnvs
+	// means no filter, and any process can be inserted the cache
 	filteredEnvs []string
+
+	in      chan *process
+	stopped chan struct{}
+	stop    sync.Once
+
+	stats struct {
+		cacheEvicts   uint64 `stats:"atomic"`
```
[the remark this answers is not in the document]
Comment (hmahmood) [reply pkg/network/tracer/process_cache.go:62]: > The first word in a variable or in an allocated struct, array, or slice can be relied upon to be 64-bit aligned.

So it should be aligned correctly.

[the remark this answers is not in the document]
Comment (vickenty) [reply pkg/network/tracer/process_cache.go:62]: But `stats` is not allocated, but embedded in another struct? [This program](https://godbolt.org/z/8EGfPndK7) prints 12, which seems to suggest that this isn't enough to ensure 64-bit alignment.

[the remark this answers is not in the document]
Comment (vickenty) [reply pkg/network/tracer/process_cache.go]: Another option is https://pkg.go.dev/go.uber.org/atomic; it used in a number of places in agent already.

[the remark this answers is not in the document]
Comment (hmahmood) [reply pkg/network/tracer/process_cache.go]: The types in that package would have the same issue if not "allocated". See https://github.com/open-telemetry/opentelemetry-collector/issues/5200#issuecomment-1098074422

File: pkg/network/tracer/process_cache.go [hunk 3]
```diff
@@ -0,0 +1,299 @@
...
+}
+
+func (pc *processCache) processEvent(entry *smodel.ProcessCacheEntry) *process {
+	var envs map[string]string
+	if entry.EnvsEntry != nil {
+		for _, v := range entry.EnvsEntry.Values {
+			kv := strings.SplitN(v, "=", 2)
+			k := kv[0]
+			found := len(pc.filteredEnvs) == 0
+			for _, r := range pc.filteredEnvs {
```
Comment (mbotarro) [inline pkg/network/tracer/process_cache.go:149]: Would it be better to use a map here instead of iterating over the `filteredEvents` slice?

File: pkg/network/tracer/process_cache.go [hunk 4]
```diff
@@ -0,0 +1,299 @@
...
+				envs = make(map[string]string)
+			}
+			envs[k] = v
+			if len(pc.filteredEnvs) > 0 && len(pc.filteredEnvs) == len(envs) {
+				break
+			}
+		}
+	}
+
+	if len(envs) == 0 && len(pc.filteredEnvs) > 0 && entry.ContainerID == "" {
```
Comment (mbotarro) [inline pkg/network/tracer/process_cache.go:174]: Shouldn't we return `nil` if `len(env)` < `len(pc.filteredEvns)` or the process must not match all `filteredEnvs` tags?

File: pkg/security/config/config.go [hunk 5]
```diff
@@ -151,8 +151,10 @@ type Config struct {
 	RuntimeCompiledConstantsEnabled bool
 	// RuntimeCompiledConstantsIsSet is set if the runtime compiled constants option is user-set
 	RuntimeCompiledConstantsIsSet bool
-	// EventMonitoring enables event monitoring
-	EventMonitoring bool
+	// NetworkProcessEventMonitoringEnabled is set to true if `network_config.enable_process_event_monitoring` is set
+	NetworkProcessEventMonitoringEnabled bool
+	// ProcessEventMonitoringEnabled is set to true if ``
```
Comment (mbotarro) [inline pkg/security/config/config.go:156]:
```suggestion
        // ProcessEventMonitoringEnabled is set to true if process lifecycle events is enabled
```

**Review by mbotarro — APPROVED** [review]
✅  for the `runtime_security_config.event_monitoring` split.

File: pkg/network/tracer/process_cache.go (hunk 4 above)
Comment (hmahmood) [reply pkg/network/tracer/process_cache.go:174]: The process doesn't have to match all the `filteredEnvs`.

File: pkg/security/config/config.go [hunk 6]
```diff
@@ -220,9 +223,11 @@ func NewConfig(cfg *config.Config) (*Config, error) {
 		EventStreamBufferSize:              coreconfig.Datadog.GetInt("runtime_security_config.event_stream.buffer_size"),
 
 		// runtime compilation
-		RuntimeCompilationEnabled:       coreconfig.Datadog.GetBool("runtime_security_config.runtime_compilation.enabled"),
-		RuntimeCompiledConstantsEnabled: coreconfig.Datadog.GetBool("runtime_security_config.runtime_compilation.compiled_constants_enabled"),
-		RuntimeCompiledConstantsIsSet:   coreconfig.Datadog.IsSet("runtime_security_config.runtime_compilation.compiled_constants_enabled"),
+		RuntimeCompilationEnabled:            coreconfig.Datadog.GetBool("runtime_security_config.runtime_compilation.enabled"),
+		RuntimeCompiledConstantsEnabled:      coreconfig.Datadog.GetBool("runtime_security_config.runtime_compilation.compiled_constants_enabled"),
+		RuntimeCompiledConstantsIsSet:        coreconfig.Datadog.IsSet("runtime_security_config.runtime_compilation.compiled_constants_enabled"),
+		NetworkProcessEventMonitoringEnabled: coreconfig.Datadog.GetBool("runtime_security_config.event_monitoring.network_process.enabled"),
```
Comment (safchain) [inline pkg/security/config/config.go:229]: just wondering if should move this to a more common key instead of using the `runtime_security_config` key ? like `system_probe_config.event_monitoring` or just `event_monitoring` ?

**Review by safchain — APPROVED** [review]
a small question regarding the config params otherwise LGTM

**Comment by brycekahle** [comment]
For visibility, noting that this change requires kernel version 4.10 because CWS uses LRU maps.

File: pkg/security/config/config.go (hunk 6 above)
Comment (hmahmood) [reply pkg/security/config/config.go]: Changed them to be under `event_monitoring_config`, e.g. `event_monitoring_config.network_process.enabled`

**Review by djmitche — APPROVED** [review]
For pkg/config/config.go

File: pkg/network/event_common.go [hunk 7]
```diff
@@ -232,10 +232,14 @@ type ConnectionStats struct {
 	Family           ConnectionFamily
 	Direction        ConnectionDirection
 	SPortIsEphemeral EphemeralPortType
-	Tags             uint64
+	StaticTags       uint64
+	Tags             map[string]interface{}
```
Comment (brycekahle) [inline pkg/network/event_common.go:236]: `map[string]struct{}` is the idiomatic way of doing a set. The Go compiler understands the empty struct and will not use any storage for the value.

File: pkg/network/tracer/process_cache.go [hunk 8]
```diff
@@ -0,0 +1,300 @@
...
+	// with the same pid but differing start times up to a max of
+	// maxProcessListSize. this is used to determine the closest
+	// match to a connection's tiimestamp
+	cacheByPid map[uint32]processList
+	// lru cache; keyed by (pid, start time)
+	cache *lru.Cache
+	// filteredEnvs contains environment variable names
+	// that a process in the cache must have; empty filteredEnvs
+	// means no filter, and any process can be inserted the cache
+	filteredEnvs map[string]interface{}
```
Comment (brycekahle) [inline pkg/network/tracer/process_cache.go:63]: `map[string]struct{}`

File: pkg/network/tracer/tracer.go [hunk 9]
```diff
@@ -315,11 +338,55 @@ func (t *Tracer) storeClosedConnections(connections []network.ConnectionStats) {
 	t.state.StoreClosedConnections(connections)
 }
 
+func monotonicToWallTime(t uint64) (uint64, error) {
```
Comment (brycekahle) [inline pkg/network/tracer/tracer.go:341]: Can you look at https://github.com/DataDog/datadog-agent/pull/12475 and make sure this function is still correct.

File: pkg/process/net/resolver/resolver.go [hunk 10]
```diff
@@ -85,11 +85,24 @@ func (l *LocalResolver) Resolve(c *model.Connections) {
...
+
 		// resolve laddr
-		cid, ok := l.ctrForPid[int(conn.Pid)]
-		if !ok {
+		//
+		// if process monitoring is enabled in the saystem-probe,
+		// then laddr container id may be set, so check that
+		// first
+		var cid string
+		if cid = conn.Laddr.ContainerId; cid == "" {
```
Comment (brycekahle) [inline pkg/process/net/resolver/resolver.go:98]:
```suggestion
		cid := conn.Laddr.ContainerId
		if cid == "" {
```

File: pkg/security/config/config.go [hunk 11]
```diff
@@ -224,9 +227,11 @@ func NewConfig(cfg *config.Config) (*Config, error) {
 		EnvsWithValue:                      coreconfig.Datadog.GetStringSlice("runtime_security_config.envs_with_value"),
 
 		// runtime compilation
-		RuntimeCompilationEnabled:       coreconfig.Datadog.GetBool("runtime_security_config.runtime_compilation.enabled"),
-		RuntimeCompiledConstantsEnabled: coreconfig.Datadog.GetBool("runtime_security_config.runtime_compilation.compiled_constants_enabled"),
-		RuntimeCompiledConstantsIsSet:   coreconfig.Datadog.IsSet("runtime_security_config.runtime_compilation.compiled_constants_enabled"),
+		RuntimeCompilationEnabled:            coreconfig.Datadog.GetBool("runtime_security_config.runtime_compilation.enabled"),
+		RuntimeCompiledConstantsEnabled:      coreconfig.Datadog.GetBool("runtime_security_config.runtime_compilation.compiled_constants_enabled"),
+		RuntimeCompiledConstantsIsSet:        coreconfig.Datadog.IsSet("runtime_security_config.runtime_compilation.compiled_constants_enabled"),
+		NetworkProcessEventMonitoringEnabled: coreconfig.Datadog.GetBool("event_monitoring_config.network_process.enabled"),
```
Comment (brycekahle) [inline pkg/security/config/config.go:233]: nit: add empty line because these are not related to runtime compilation
```suggestion
		
		NetworkProcessEventMonitoringEnabled: coreconfig.Datadog.GetBool("event_monitoring_config.network_process.enabled"),
```

File: pkg/network/tracer/process_cache.go [hunk 12]
```diff
@@ -0,0 +1,300 @@
...
+		if ts < pl[i].StartTime {
+			continue
+		}
+
+		if closest == nil {
+			closest = pl[i]
+			continue
+		}
+
+		if pl[i].StartTime > closest.StartTime {
```
Comment (brycekahle) [inline pkg/network/tracer/process_cache.go:294]: This seems to be choosing the latest process, rather than the one closest to the timestamp?

File: pkg/network/tracer/process_cache.go [hunk 13]
```diff
@@ -0,0 +1,300 @@
...
+	for pid, pl := range pc.cacheByPid {
+		res[pid] = pl
+	}
+
+	return res, nil
+}
+
+func (pl processList) update(p *process) processList {
+	for i := range pl {
+		if pl[i].Pid == p.Pid && pl[i].StartTime == p.StartTime {
```
Comment (brycekahle) [inline pkg/network/tracer/process_cache.go:259]: The `processList` is already filtered by PID right?

File: pkg/network/tracer/tracer.go (hunk 9 above)
Comment (hmahmood) [reply pkg/network/tracer/tracer.go]: I copied over `TimeResolver` from `pkg/security/probe` for now, but I'm going to move that to `pkg/util` in a separate PR.

File: pkg/network/tracer/process_cache.go (hunk 13 above)
Comment (hmahmood) [reply pkg/network/tracer/process_cache.go]: Right, fixed

File: pkg/network/tracer/process_cache.go (hunk 12 above)
Comment (hmahmood) [reply pkg/network/tracer/process_cache.go]: Fixed

File: cmd/system-probe/config/config_linux_test.go [hunk 14]
```diff
@@ -0,0 +1,42 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
```
Comment (guyarb) [inline cmd/system-probe/config/config_linux_test.go:1]: if that's a linux file (by the name)
shouldn't we use `linux_bpf` build tag?

File: pkg/network/tracer/process_cache.go [hunk 15]
```diff
@@ -0,0 +1,313 @@
...
+	smodel "github.com/DataDog/datadog-agent/pkg/security/secl/model"
+	"github.com/DataDog/datadog-agent/pkg/util/atomicstats"
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+)
+
+var defaultFilteredEnvs = []string{
+	"DD_ENV",
+	"DD_VERSION",
+	"DD_SERVICE",
+}
```
Comment (guyarb) [inline pkg/network/tracer/process_cache.go:27]: shouldn't we make that configurable to allow flexibility?

**Review by guyarb — APPROVED** [review]
LGTM
left 2 questions to consider

File: cmd/system-probe/config/config_linux_test.go (hunk 14 above)
Comment (hmahmood) [reply cmd/system-probe/config/config_linux_test.go:1]: I guess for the tracer linux with ebpf dependency is true (which would imply `linux_bpf`), but this file is strictly linux (no bpf dependency).

**Review by ISauve — APPROVED** [review]

File: pkg/process/net/resolver/resolver.go [hunk 16]
```diff
@@ -85,11 +85,24 @@ func (l *LocalResolver) Resolve(c *model.Connections) {
...
 
+		if conn.Laddr == nil {
+			continue
+		}
+
 		// resolve laddr
-		cid, ok := l.ctrForPid[int(conn.Pid)]
-		if !ok {
+		//
+		// if process monitoring is enabled in the saystem-probe,
```
Comment (akarpz) [inline pkg/process/net/resolver/resolver.go:94]: nit: `saystem`

**Review by derekwbrown — APPROVED** [review]

**Review by akarpz — APPROVED** [review]

**Review by sgnn7 — APPROVED** [review]
LGTM for ASC-owned `pkg/config/config.go`
